tools/labs/skels/device_drivers/kernel/so2_cdev.c

```diff
@@ -0,0 +1,219 @@
+/*
+ * Character device drivers lab
+ *
+ * All tasks
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/fs.h>
+#include <linux/cdev.h>
+#include <linux/uaccess.h>
+#include <linux/sched.h>
+#include <linux/wait.h>
+
+#include "../include/so2_cdev.h"
+
+MODULE_DESCRIPTION("SO2 character device");
+MODULE_AUTHOR("SO2");
+MODULE_LICENSE("GPL");
+
+#define LOG_LEVEL	KERN_INFO
+
+#define MY_MAJOR		42
+#define MY_MINOR		0
+#define NUM_MINORS		1
+#define MODULE_NAME		"so2_cdev"
+#define MESSAGE			"hello\n"
+#define IOCTL_MESSAGE		"Hello ioctl"
+
+#ifndef BUFSIZ
+#define BUFSIZ		4096
+#endif
+
+
+struct so2_device_data {
+	/* add cdev member */
+	struct cdev cdev;
+	/* add buffer with BUFSIZ elements */
+	char buffer[BUFSIZ];
+	size_t size;
+	/* TODO 7: extra members for home */
+	/* add atomic_t access variable to keep track if file is opened */
+	atomic_t access;
+};
+
+struct so2_device_data devs[NUM_MINORS];
+
+static int so2_cdev_open(struct inode *inode, struct file *file)
+{
+	struct so2_device_data *data;
+
+	/* print message when the device file is open. */
+	printk("deviced opened");
+
+	/* inode->i_cdev contains our cdev struct, use container_of to obtain a pointer to so2_device_data */
+	data = container_of(inode->i_cdev, struct so2_device_data, cdev);
+
+	file->private_data = data;
+
+	/* return immediately if access is != 0, use atomic_cmpxchg */
+	if (atomic_cmpxchg(&data->access, 0, 1) != 0)
+	{
+		return -EBUSY;
+	}
+
+	set_current_state(TASK_INTERRUPTIBLE);
+	schedule_timeout(10 * HZ);
+
+	return 0;
+}
+
+static int
+so2_cdev_release(struct inode *inode, struct file *file)
+{
+	/* print message when the device file is closed. */
+	printk("device closed");
+
+#ifndef EXTRA
+	struct so2_device_data *data =
+		(struct so2_device_data *) file->private_data;
+
+	/* reset access variable to 0, use atomic_set */
+	atomic_set(&data->access, 0);
+#endif
+	return 0;
+}
+
+static ssize_t
+so2_cdev_read(struct file *file,
+		char __user *user_buffer,
+		size_t size, loff_t *offset)
+{
+	struct so2_device_data *data =
+		(struct so2_device_data *) file->private_data;
+	size_t to_read;
+
+#ifdef EXTRA
+	/* TODO 7: extra tasks for home */
+#endif
+
+	/* Copy data->buffer to user_buffer, use copy_to_user */
+	to_read = min(data->size - *offset, size);
+	
+	if (to_read <= 0)
+	{
+        return 0;
+	}
+	
+	if (copy_to_user(user_buffer, data->buffer + *offset, to_read) != 0)
+	{
+		return -EFAULT;
+	}
+	*offset += to_read;
+
+	return to_read;
+}
+
+static ssize_t
+so2_cdev_write(struct file *file,
+		const char __user *user_buffer,
+		size_t size, loff_t *offset)
+{
+	struct so2_device_data *data =
+		(struct so2_device_data *) file->private_data;
+
+
+	/* copy user_buffer to data->buffer, use copy_from_user */
+	if (copy_from_user(data->buffer + *offset, user_buffer, size) != 0)
```

Comment:
If size is biggger than the remaining space in the buffer, read only the amount that is available